### DISCUSSION TITLE
Can the framework handle multiple files?

### DISCUSSION THREAD

**Discussion by pascal-nil** [rwf2/Rocket#2634] [Questions] [answered]
**Description**
I'm starting to learn Rust, so maybe I missed an important notion, but I can't find anything in the doc, and no examples. 
I need to receive several files in a single request. 
In Golang, I can receive a slice of my files and then loop over them, but I can't find the same thing in this library. 
Here's the function in Golang : 
```
files := r.MultipartForm.File["files"]
for _, fh := range files {/* ... */}
```
Docs from Golang : [https://pkg.go.dev/mime/multipart](doc)
Here's what my input file sends in a ```<form>```: 
```
<input type="file" name="files"  multiple />
```
Here's my code in Rust: 

```
#[derive(FromForm)]
struct Upload<'r> {
    save: bool,
    file: TempFile<'r>,
}

#[post("/upload", data = "<upload>")]
fn upload(upload: Form<Upload<'_>>) { /* .. */ }
```

How can I loop with multiple file input ? 

**Environment:**

 - Rocket Version:latest (0.5.0-rc.3)

Thanks

**Comment by SergioBenitez** [comment, marked as answer]
Of course. Here's a fully working application:

```rust
use rocket::fs::TempFile;
use rocket::response::content::RawHtml;
use rocket::form::Form;

#[get("/")]
fn upload_page() -> RawHtml<&'static str> {
    RawHtml(r#"
        <!DOCTYPE html>
        <html lang="en">
            <head>
                <title>Upload Example</title>
            </head>
            <body>
                <form method="post" action="/upload" enctype="multipart/form-data">
                    <label for="files">Select files:</label>
                    <input type="file" id="files" name="[]" multiple><br><br>
                    <input type="submit">
                </form>
            </body>
        </html>
    "#)
}

#[post("/", data = "<files>")]
fn upload_post(files: Form<Vec<TempFile<'_>>>) -> String {
    let files = files.into_inner();
    for (i, file) in files.iter().enumerate() {
        println!("{i}: {}", file.len());
    }

    format!("{} files received", files.len())
}

#[launch]
fn rocket() -> _ {
    rocket::build()
        .mount("/upload", routes![upload_page, upload_post])
}
```

You probably want to actually name the field something, in which case you'd use a struct:

```rust
#[derive(FromForm)]
struct Upload<'r> {
    files: Vec<TempFile<'r>>,
}
```

Then you can name the form field `files` or `files[]`.

It doesn't have to use `TempFile`. You could choose to always store in memory by using an `&[u8]` instead, for example, or create your own form field type that streams the data somewhere else.

**Comment by SergioBenitez** [comment]
> so maybe I missed an important notion, but I can't find anything in the doc, and no examples.

We can't possibly enumerate every possibility for forms, but the forms guide is rather comprehensive: https://rocket.rs/v0.5-rc/guide/requests/#forms, and the forms example illustrated several uses of `Vec`.
